### ISSUE TITLE
Allow Markdown configuration for Mattermost and Slack reporters

### ISSUE THREAD

**Issue by jamesquilty** [thp/urlwatch#597]
I have configured urlwatch to report to an incoming webhook on my workplace's Mattermost server using the Slack reporter (#596). At present it appears that the reporter allows only text configuration and because this text is interpreted as Markdown by Mattermost there are several unintended formatting artefacts in the formatted post which frequently render the urlwatch output &mdash; particularly diff output &mdash; unreadable. It would be helpful if Markdown could be added as a configuration option for the Mattermost and Slack reporters, as it is for the Matrix reporter.

**Comment by scottmac** [comment]
You can use a filter on the URL turn the content into markdown

```
filter:
- css: ul#monitor
- html2text:
    method: pyhtml2text
    ul_item_mark: '-'
    strong_mark: '*'
```

**Comment by thp** [comment]
Patches to improve this gladly accepted. Alternatively, maybe just a documentation of the filters in "Advanced" section could help?

Alternatively, there's also https://github.com/thp/urlwatch/pull/572 which improves reporting with Markdown for Matrix, maybe this can be repurposed so that Slack + Mattermost use the `MarkdownReporter` instead of the `TextReporter`?

**Comment by jamesquilty** [comment]
@scottmac That's good to know for the input side of things!

@thp I'll take a look to see if a patch would be within my Python skills, but those skills are rudimentary so I'm not hopeful. I do appreciate your attention on this bug.

An example of what's going on might help, as there are a couple of things that are problematic. The lines starting with `- ` and `+ ` (note the space) that are interpreted as a list by Markdown renderers. The separating text `---------------------------------------------------------------------------` is rendered as a horizontal rule when preceded by a blank line and as H1 or H2 heading (depending on the renderer) when preceded by text. Thus, 

```text
---------------------------------------------------------------------------
CHANGED: Title ( URL )
---------------------------------------------------------------------------
--- @	Wed, 02 Dec 2020 16:00:01 +1300
+++ @	Tue, 08 Dec 2020 16:00:02 +1300
@@ -13,7 +13,7 @@
    Unchanged document (328.9 KB)
   </a>
   <a data-assetid="1798728" href="https://www.example.com/etc/etc/some-document.pdf" id="d5">
-   Previous document (362.4 KB)
+   Current document (108.0 KB)
   </a>
  </body>
 </html>
```

renders as:

---------------------------------------------------------------------------
CHANGED: Title ( URL )
---------------------------------------------------------------------------
--- @	Wed, 02 Dec 2020 16:00:01 +1300
+++ @	Tue, 08 Dec 2020 16:00:02 +1300
@@ -13,7 +13,7 @@
    Unchanged document (328.9 KB)
   </a>
   <a data-assetid="1798728" href="https://www.example.com/etc/etc/some-document.pdf" id="d5">
-   Previous document (362.4 KB)
+   Current document (108.0 KB)
   </a>
  </body>
 </html>

or, something like that depending on the renderer.

I _think_ that this is already handled by (﻿﻿MarkdownReporter(ReporterBase))[https://github.com/thp/urlwatch/blob/master/lib/urlwatch/reporters.py#L726]?

**Comment by dkasak** [comment]
@jamesquilty Yes, this is already handled by the `MarkdownReporter`. The trick is to enclose the diff into a Markdown fenced code block.

````
```
Diff lines
Go here
```
````

**Comment by knutwannheden** [comment]
AFAICT Slack doesn't support the language as indicated by Markdown code blocks :-( Also using the Slack desktop application the only way I know of is to add a "text snippet" to the message, where syntax highlighting can be specified. I took a quick look at the documentation, but I couldn't see how an incoming webhook can be used to include a "text snippet".

**Comment by knutwannheden** [comment]
It seems like the Slack reporter would have to upload the diff as a file to get the diff rendering with colors. That is then however not supported by the incoming webhooks and the reporter would instead have to use a REST API with an OAuth token.

I will probably look into implementing a custom reporter for myself. If there is interest in this I could consider creating a PR for this.

**Comment by jamesquilty** [comment]
It seems that Mattermost and Slack have divergent capabilities. Perhaps this is a case for splitting the urlwatch reporters?

**Comment by knutwannheden** [comment]
I implemented my own reporter now which uploads files instead, which allows colored rendering of diffs in Slack. A drawback is that this doesn't work on Android (no idea about iOS), but good enough for me.

**Comment by jamesquilty** [comment]
Picking this up again. My expectation is that the configuration below would cause the payload sent to the Mattermost webhook to be changed by `MarkdownReporter` _but_ what is received by the webhook is identical to the plain text sent to `stdout` (example in my comment above). That's  a bit confusing.

@dkasak putting the diff into a fenced code block would be one step toward fixing the problem... how would this be accomplished? Is there a configuration setting, or do I have to use `diff_filter` with a custom script, or is modification of the Python code required?

```yaml
display:
  error: true
  new: true
  unchanged: false
job_defaults:
  all: {}
  browser: {}
  shell: {}
  url: {}
report:
  ...
  mattermost:
    enabled: true
    webhook_url: '...'
    markdown: true
  stdout:
    color: true
    enabled: true
  markdown:
    enabled: true
    details: true
    minimal: true
    footer: false
```
